programs/clearing_house/src/controller/position.rs

```diff
@@ -69,6 +70,264 @@ pub struct PositionDelta {
+                        .quote_entry_amount
+                        .checked_mul(delta_base_asset_amount_unsigned)
+                        .ok_or_else(math_error!())?
+                        .checked_div(base_asset_amount_before_unsigned)
+                        .ok_or_else(math_error!())?;
+
+                    let quote_entry_amount = position
+                        .quote_entry_amount
+                        .checked_sub(entry_amount_delta)
+                        .ok_or_else(math_error!())?;
+
+                    (
+                        position
+                            .quote_asset_amount
+                            .checked_sub(cost_basis)
+                            .ok_or_else(math_error!())?,
+                        quote_entry_amount,
+                    )
+                };
+
+            (
+                new_quote_asset_amount,
+                new_quote_entry_amount,
+                new_base_asset_amount,
+                pnl,
+            )
+        } else {
+            let new_quote_asset_amount = position
+                .quote_asset_amount
+                .checked_add(delta.quote_asset_amount)
+                .ok_or_else(math_error!())?;
+            let new_quote_entry_amount = position
+                .quote_entry_amount
+                .checked_add(delta.quote_asset_amount)
+                .ok_or_else(math_error!())?;
+            let new_base_asset_amount = position
+                .base_asset_amount
+                .checked_add(delta.base_asset_amount)
+                .ok_or_else(math_error!())?;
+
+            (
+                new_quote_asset_amount,
+                new_quote_entry_amount,
+                new_base_asset_amount,
+                0_i128,
+            )
+        };
+
+    // let _reduced_position = !increasing_position
+    //     && position.base_asset_amount.signum() == new_base_asset_amount.signum();
+    let closed_position = new_base_asset_amount == 0;
+    let flipped_position = position.base_asset_amount.signum() != new_base_asset_amount.signum();
+
+    // Update user position
+    if closed_position {
+        position.last_cumulative_funding_rate = 0;
+        position.last_funding_rate_ts = 0;
+    } else if new_position || flipped_position {
+        if new_base_asset_amount > 0 {
+            position.last_cumulative_funding_rate = market.amm.cumulative_funding_rate_long;
+        } else {
+            position.last_cumulative_funding_rate = market.amm.cumulative_funding_rate_short;
+        }
+    }
+
+    position.quote_asset_amount = new_quote_asset_amount;
+    position.quote_entry_amount = new_quote_entry_amount;
+    position.base_asset_amount = new_base_asset_amount;
+
+    market.amm.market_position_per_lp = position;
+
+    Ok(pnl)
+}
+
+pub fn update_amm_position(
+    market: &mut Market,
+    delta: &PositionDelta,
+) -> ClearingHouseResult<i128> {
+    let mut position = market.amm.market_position;
```

Comment:
This is just a duplicate of `update_position_and_market` right? Whats the plan to make them go through the same code path?

Reply:
resolved fyi. made helper function `calculate_position_new_quote_base_pnl` (flex on name)